client/normal_value.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package client
+
+import (
+	"time"
+
+	"github.com/sourcenetwork/immutable"
+)
+
+// NormalValue is the interface for the normal value types.
+// It is used to represent the normal (or standard) values across the system and to avoid
+// asserting all possible types like int, int32, int64, etc.
+//
+// All methods returning a specific type returns the value and the second boolean flag indicating
+// if the value is of the requested type.
+//
+// All nillable values are represented as immutable.Option[T].
+type NormalValue interface {
+	// Any returns the underlying value.
+	// If the value is nillable the result will be of type `immutable.Option[T]`.
+	Any() any
+	// Unwrap returns the underlying value.
+	// For not nillable values it will act as `Any()`
+	// For nillable values it will return result of `Value()` of `immutable.Option` if it
+	// has value, otherwise it will return `nil`.
+	Unwrap() any
+
+	// IsNil returns if the value is nil.
+	IsNil() bool
+	// IsNillable returns if the value can be nil.
+	IsNillable() bool
+
+	// Bool returns the value as a bool.
+	Bool() (bool, bool)
+	// Int returns the value as an int64.
+	Int() (int64, bool)
+	// Float returns the value as a float64.
+	Float() (float64, bool)
+	// String returns the value as a string.
+	String() (string, bool)
+	// Bytes returns the value as a byte slice.
+	Bytes() ([]byte, bool)
+	// Time returns the value as a time.Time.
+	Time() (time.Time, bool)
+	// Document returns the value as a *Document.
+	Document() (*Document, bool)
+
+	// NillableBool returns the value as a nillable bool.
```

Comment:
there is actually a note in the documentation of the interface itself. I thought it would be enough.

Reply:
Ah true, I know it is a hassle, but can you please document it on the funcs too - I think that is best practice, and it is the first place where people will look most of the time to try and find that out.

Reply:
I think the time spent now will be much smaller than the sum of all the head scratching (or bugs) that could take place if people have to chase down this info (or guess and potentially get it wrong)

Reply:
> document whether they return nillable values for none nillable `NormalValues`

I don't think we should document on what it returns in case it's nillable or not nillable. It's irrelevant. If the requested method does not match exactly the underlying type, the second param will be false (even if both underlying and requested are nillable). And in this case the first parameter is the default value as with standard Go type assertion. We don't document that the first value of `Int()` returns 0, because it should be clear and nobody is supposed to use this value anyway.

Reply:
> I don't think we should document on what it returns in case it's nillable or not nillable. It's irrelevant. If the requested method does not match exactly the underlying type, the second param will be false (even if both underlying and requested are nillable).

This is not clear from the function signature.  It is not unreasonable to expect `NillableInt` to return Some when calling it on a none nillable int type.

Reply:
done